### PR TITLE
Introduce async/await

### PR DESCRIPTION
**Pull request by pietbrauer**

### CODE REVIEW

File: Sources/RequestKit/JSONPostRouter.swift [hunk 1]
```diff
@@ -5,7 +5,13 @@ import FoundationNetworking
...
 public protocol JSONPostRouter: Router {
     func postJSON<T>(_ session: RequestKitURLSession, expectedResultType: T.Type, completion: @escaping (_ json: T?, _ error: Error?) -> Void) -> URLSessionDataTaskProtocol?
-    func post<T: Codable>(_ session: RequestKitURLSession, decoder:JSONDecoder, expectedResultType: T.Type, completion: @escaping (_ json: T?, _ error: Error?) -> Void) -> URLSessionDataTaskProtocol?
+    func post<T: Codable>(_ session: RequestKitURLSession, decoder: JSONDecoder, expectedResultType: T.Type, completion: @escaping (_ json: T?, _ error: Error?) -> Void) -> URLSessionDataTaskProtocol?
+    
+    @available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+    func postJSON<T>(_ session: RequestKitURLSession, expectedResultType: T.Type) async throws -> T?
+    
+    @available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+    func post<T: Codable>(_ session: RequestKitURLSession, decoder: JSONDecoder, expectedResultType: T.Type) async throws -> T?
```
Comment (417-72KI) [inline Sources/RequestKit/JSONPostRouter.swift]: I think these `async` functions should be enclosed with `#if swift(>=5.5) ~ #endif` to be ignored by non-latest Xcodes 🤔

Comment (pietbrauer) [reply Sources/RequestKit/JSONPostRouter.swift]: Doesn’t seem to work. Also `compiler(>=5.5)` seems to still trigger under Xcode 12.4

**Comment by pietbrauer** [comment]
This is kind of a mess:
- [x] GitHub Actions doesn't include Xcode 13 (yet) so tvOS, iOS and watchOS fail
- [x] macOS Monterey is not released and so async/await is not on macOS (yet)
- [x] Linux (FoundationNetworking) doesn't have async/await at all and this needs implementation by someone
- [x] I will exclude async/await functionality when using `FoundationNetworking` for now

Really struggling on what to do now with this PR.

**Comment by pkandrashkou** [comment]
Hello @pietbrauer. I think it would have been better to release these changes as a major version bump as of now it breaks some projects such as https://github.com/nerdishbynature/octokit.swift and https://github.com/danger/swift. 

I could not find any issues related to this yet, but I can see that the release is still quite fresh, so I assume this is not only my local issue.

I think the main reason is the deletion of the `Response` structure [here](https://github.com/nerdishbynature/RequestKit/pull/75/files#diff-fd1f744e6d6361b564033fe24980d6b9fae7938e5e054ebb3afe196c12fd26aeL6). However, I am not 100% sure this is the only breaking change because I have not investigated it further and as far as I can see there are quite a lot of changes apart from that one.

Nevertheless, I think another part of the story is that these projects use `RequestKit` as a dependency with `package(name:url:from:)` and the [Apple Documentation](https://developer.apple.com/documentation/swift_packages/package/dependency/3582850-package) states:
> Adds a package dependency that uses the version requirement, starting with the given minimum version, going up to the next major version.

Personally, I think it would be better to revert and re-release the `RequestKit` to save the trouble for other developers changing their `Package.swift` files to pin the `2.3.0` release or updating their the usage code to the new API which is more time-consuming tbh 🙂

Please let me know if more information is required from my side. I would be glad to help.

**Pull request merged**